Delete process instance related data from ES and OS

remcowesterhoud opened the pull request:
## Description

<!-- Describe the goal and purpose of this PR. -->

This PR adds the implementation to the `HistoryDeletionJob`. The job already queried the PIs that need to be deleted. With this PR we perform the actual deletion. For a process instance this results in 3 requests to ES per batch. The first request will delete all the PI related data from the leaves (variables, messages, etc). The second request will remove the PI itself from the list view. The final request will remove the PIs from the history-deletion index. At this point the deletion is finalized.

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

closes #41105 
closes #41109

remcowesterhoud committed:
refactor: retrieve the deletion entities as list

The structure we had was not very useable. It's easier to just return the full entities. We need more data than only
the id anyways.

remcowesterhoud committed:
feat: delete PI related data from Elasticsearch

remcowesterhoud committed:
feat: delete PI related data from list-view and from the history-deletion index

The list view is not a process instance dependant, instead this is the "root" of the process instance in ES/OS. Since
 we have deleted all the leaves, we can safely delete the roots afterwards.

 Once all the roots also succeeded in deletion we can delete the process instances in the batch from the
 history-deletion index itself.

remcowesterhoud committed:
docs: fix javadoc after refactorings

remcowesterhoud committed:
test: verify history deletion job deletes process instances

remcowesterhoud committed:
test: verify history deletion repostiory deletes properly from ES/OS

remcowesterhoud committed:
feat: delete PI related data from OpenSearch

wasup-yash reviewed (approved):
LGTM

konstantinos-mylosis reviewed (approved):
LGTM, just curious
❓Is there a way to have a common scheduled job for camunda-rdbms exporters + interface + different impl based on config.connect().getTypeEnum()? -- `DatabaseType`

houssain-barouni commented on zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/historydeletion/HistoryDeletionJob.java:
```diff
@@ -30,31 +34,87 @@ public class HistoryDeletionJob implements BackgroundTask {
+    return CompletableFuture.allOf(deletionFutures.toArray(CompletableFuture[]::new))
+        .thenCompose(
+            ignored -> {
+              final var dependentSourceIdx = listViewTemplate.getFullQualifiedName();
```
❓ This will delete the process instance for the runtime list-view index only and not for the historical indices (dated indices), is it expected or will it be done somewhere else?

remcowesterhoud replied:
@houssain-barouni good catch, it must be deleted from the dated indices as well. I'll make sure to add the `*` 🙂

remcowesterhoud replied:
And add it to the tests, cause I missed this 🤦

houssain-barouni replied:
you can use `indexTemplateDescriptor.getIndexPattern()`

houssain-barouni reviewed (commented):
I had one question whether the deletion should be done in the historical indices as well?

remcowesterhoud committed:
fix: delete data from dated list-view indices

The process instances must be removed from the dated list-view indices as well.
